cmd/jujud/updateseries/updateseries.go

```diff
@@ -118,38 +139,64 @@ func (c *UpdateSeriesCommand) Run(ctx *cmd.Context) error {
 		return err
 	}
 
+	jujuVersion, err := GetJujuVersion(c.machineAgent, c.dataDir)
+	if err != nil {
+		return err
+	}
+
 	switch {
 	case toInitSys == service.InitSystemUpstart:
 		return errors.NewNotSupported(nil, "downgrade to series using upstart not supported")
 	case toInitSys == fromInitSys && toInitSys == service.InitSystemSystemd:
-		if err = c.copyTools(); err != nil {
+		if err := c.manager.CopyAgentBinary(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries, c.fromSeries, jujuVersion); err != nil {
 			return err
 		} else {
 			ctx.Infof("successfully copied tools and relinked agent tools")
 		}
 		if !c.startAgents {
 			break
 		}
-		if err = c.startAllAgents(ctx); err != nil {
+		if errMachineName, errUnitNames, err := c.manager.StartAllAgents(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries); err != nil {
 			return err
+		} else {
+			for _, unitName := range errUnitNames {
+				ctx.Infof("started %s service", unitName)
+			}
+			ctx.Infof("started %s service", errMachineName)
 		}
 		ctx.Infof("all agents successfully restarted")
 	case toInitSys == service.InitSystemSystemd:
 		errorHappened := false
-		if err = c.writeSystemdAgents(ctx); err != nil {
+		errAgentNames = errAgentNames[:0]
+		if errSysdServiceNames, errSymServiceNames, errAgentNames, err = c.manager.WriteSystemdAgents(c.machineAgent, c.unitAgents, c.dataDir, systemdDir, systemdMultiUserDir, c.toSeries); err != nil {
 			ctx.Warningf("%s", err)
+			for _, agentName := range errAgentNames {
+				ctx.Warningf("failed to write service for %s: %s", agentName, err)
```

Comment:
Isn't it possible for some service writes to succeed and a different one to fail?